### PR TITLE
Gutenberg: Add Document Title to the Gutenberg Editor

### LINKED ISSUE #27300: Gutenberg: Add missing document title for the Gutenberg Editor
Let's use the `DocumentHead` component to set a page title. On a new post this should be:

`New Post < Site Title`

and after an autosave:

`Edit Post < Site Title`

screen shot 2018-09-18 at 5 30 59 pm
screen shot 2018-09-18 at 5 31 12 pm

### PR DESCRIPTION
**Pull request by Copons**
Fix #27300 

Add a document title to the Gutenberg editor based on the post type, and if the post is new or not.

This new component is heavily inspired by the [classic editor version](https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/editor-document-head/index.jsx), although it's slightly simplified, as the Gutenberg editor already has all all the needed info available.

For example:
- `QueryPostTypes` is already called by the [editor itself](https://github.com/Automattic/wp-calypso/blob/master/client/gutenberg/editor/main.jsx#L43), and there is no need to call it again.
- The `postId` presence is enough, for now, to determine if the editor is a New or Edit one, but this will likely change after https://github.com/Automattic/wp-calypso/pull/27301 is merged. I'll rebase and update when it happens.

## Testing instructions

Visit http://calypso.localhost:3000/gutenberg for the following routes and make sure the document title is the expected one.

| Route | Title |
| - | - |
| `/post/{ site }` | New Post |
| `/post/{ site }/{ postId }` | Edit Post |
| `/page/{ site }` | New Page |
| `/page/{ site }/{ postId }` | Edit Page |
| `/edit/{ customPostType }/{ site }` | { customPostType.labels.new_item } (e.g. New Post) |
| `/edit/{ customPostType }/{ site }/{ postId }` | { customPostType.labels.edit_item } (e.g. Edit Post) |

Note that when loading a CPT, the document title should "flash" from `New` (or `Edit`) to `New CPT` (or `Edit CPT`), and that most CPT are indeed supposed to be labelled as `Post` (e.g. `New Post` or `Edit Post`).

### CODE REVIEW

File: client/state/ui/selectors.js [hunk 1]
```diff
@@ -89,7 +89,7 @@ export function getSectionGroup( state ) {
  * @return {Boolean}       Whether current section is site-specific
  */
 export function isSiteSection( state ) {
-	return includes( [ 'sites', 'editor' ], getSectionGroup( state ) );
+	return includes( [ 'sites', 'editor', 'gutenberg' ], getSectionGroup( state ) );
```
Comment (Copons) [inline client/state/ui/selectors.js]: Note: this is needed in order to add the site title to the document title (see: https://github.com/Automattic/wp-calypso/blob/master/client/state/document-head/selectors.js#L75)

File: client/gutenberg/editor/edit-post/components/editor-document-head/index.jsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	let title = '';
+	if ( 'post' === postType ) {
+		title = isNewPost ? translate( 'New Post' ) : translate( 'Edit Post' );
+	} else if ( 'page' === postType ) {
+		title = isNewPost ? translate( 'New Page' ) : translate( 'Edit Page' );
+	} else if ( !! postTypeObject ) {
+		title = isNewPost ? postTypeObject.labels.new_item : postTypeObject.labels.edit_item;
+	} else {
+		title = isNewPost ? translate( 'New' ) : translate( 'Edit' );
+	}
```
Comment (dmsnell) [inline client/gutenberg/editor/edit-post/components/editor-document-head/index.jsx]: _if_ we split this into a separate function we can more easily test it

```js
export const getTitle = ( { isNewPost, postType, postTypeObject, translate = a => a } ) => {
	if ( 'post' === postType && isNewPost ) {
		return translate( 'New Post' );
	}

	if ( 'post' === postType && ! isNewPost ) {
		return translate( 'Edit Post' );
	}

	if ( 'page' === postType && isNewPost ) {
		return translate( 'New Page' );
	}

	if ( 'page' === postType && ! isNewPost ) {
		return translate( 'Edit Page' );
	}

	if ( !! postTypeObject && isNewPost ) {
		return postTypeObject.labels.new_item;
	}

	if ( !! postTypeObject && ! isNewPost ) {
		return postTypeObject.labels.edit_item;
	}

	if ( isNewPost ) {
		return translate( 'New' );
	}

	return translate( 'Edit' );
}

export const EditorDocumentHead = ( { isNewPost, postType, postTypeObject, translate } ) =>
	<DocumentHead title={ getTitle( { isNewPost, postType, postTypeObject, translate } ) } />
```

**Comment by Copons** [comment]
Regarding this corner case:

> The postId presence is enough, for now, to determine if the editor is a New or Edit one, but this will likely change after #27301 is merged. I'll rebase and update when it happens.

I was naively determining if the editor was "New" or "Edit" on the presence of the post ID. For this reason, I assumed that, once the PR that added the ID to the path after the first autosave was merged, it would have "broken" this logic into this behaviour:

- Open `/gutenberg/post/{ site }`, title is "New Post".
- Type something and let the editor autosave.
- Now the path is `/gutenberg/post/{ site }/{ postId }, and the title is "Edit Post", even though the post has never been published or even manually saved.

Fact is... this is what the classic editor does as well.
[`newPost`](https://github.com/Automattic/wp-calypso/blob/master/client/post-editor/editor-document-head/index.jsx#L73), the "New"/"Edit" flag, is literally based upon the [post ID](https://github.com/Automattic/wp-calypso/blob/master/client/state/ui/editor/selectors.js#L35-L37) (although through the whole editor's Redux infrastructure, that we might want to keep, but that felt overkill for this little change).

@gwwar any ideas? 🤔

File: client/gutenberg/editor/edit-post/components/editor-document-head/index.jsx (hunk 2 above)
Comment (Copons) [reply client/gutenberg/editor/edit-post/components/editor-document-head/index.jsx]: It looks so cumbersome though.
And it's not like there will be any more cases there.

**Comment by gwwar** [comment]
>  the "New"/"Edit" flag, is literally based upon the post ID

We can do this based off of draft status. A new post will have a status of "auto-draft" for like 3 seconds before autosave kicks in :)

**Comment by vindl** [comment]
Could we move this outside of `edit-post` to some  subfolder in `editor`, like we did with `BrowserURL`? It will make it easier to differentiate components that we introduced from the ones that were copied from core.

**Comment by Copons** [comment]
> We can do this based off of draft status. A new post will have a status of "auto-draft" for like 3 seconds before autosave kicks in :)

Somehow I was sure the status changed from `auto-draft` to `draft` on first autosave, but in fact it doesn't.
The status remains `auto-draft` regardless, even after publishing the post. 🤔 

Though I think that if we tied "New/Edit" to the status, this would create a different, equally incorrect, experience.
I'd expect to "edit" a post once it has been published, even if I've never left the editor.

File: client/gutenberg/editor/editor-document-head/index.jsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	postId: PropTypes.number,
+	postType: PropTypes.string,
+	postTypeObject: PropTypes.object,
+};
+
+const mapStateToProps = ( state, { postId, postType } ) => {
+	const siteId = getSelectedSiteId( state );
+	return {
+		isNewPost: ! postId,
+		postTypeObject: getPostType( state, siteId, postType ),
```
Comment (gwwar) [inline client/gutenberg/editor/editor-document-head/index.jsx]: This is the wrong post schema isn't it? We'll want to use the wp.data and the v2 posts, similar to: https://github.com/Automattic/wp-calypso/pull/27301/files#diff-16d357cdb3cf5543a9079d68a41f36d5R38

**Comment by gwwar** [comment]
I think we're also running into https://github.com/Automattic/wp-calypso/issues/27430

**Comment by gwwar** [comment]
> Somehow I was sure the status changed from auto-draft to draft on first autosave, but in fact it doesn't.

To access Gutenberg editor state see dd0d6a8. (Feel free to squish/update that as needed). If we're accessing global Calypso state, we'd need to add a query component to poll for the v1 endpoint, and it's likely why you never see the status update here.

**Comment by gwwar** [comment]
> I'd expect to "edit" a post once it has been published, even if I've never left the editor.

Ah good call, feel free to update that, you should be able to check for this using the same post status.

**Comment by Copons** [comment]
@gwwar Thanks for pushing that commit, it really helped me a lot!

I've updated this with a slightly more complicated logic that though involves `shouldComponentUpdate`, so in the end it should be an improvement overall.

Basically, the assumption is that if an editor starts as `auto-draft` it's, inevitably, a **New**.
At first autosave, the status changes into `draft`, but it's still a **New**. So, if the status changes from `auto-draft` to `draft` we don't update the component (we also don't update if the status doesn't change of course).
Now, when the status changes again, it's inevitably from `draft` to `publish`, `future`, etc.: it's now an **Edit**. Only now we allow the component to update, and the title changes from "New" to "Edit".

If we start with any non-`auto-draft` status, we assume that it's an **Edit**, and it will always be.
The title will be stuck as "Edit".

If the status is `undefined` (it is always `undefined` for before the post details are loaded), we treat it as a special case and don't add any "New/Edit" to the title.
This prevents a flash of "Edit" in the title for new posts.

Additionally, I've also applied @dmsnell [suggestion](https://github.com/Automattic/wp-calypso/pull/27346#discussion_r219238213) as it makes more sense now that this is a class component.

**Comment by kwight** [comment]
Lovely, all seems to work as expected. Noting that yes, non-posts are still a problem for the auto-draft endpoint, so published CPTs end up under "Posts". But everything here worked as expected 🎉

**Comment by kwight** [comment]
CPT and page handling should be fixed with D18864-code.

**Comment by Copons** [comment]
There is no better time like the present!
I'll ship and open a card to remind us of updating this after D18864-code lands.

**Comment by gwwar** [comment]
We should be unblocked now 👍

**Comment by Copons** [comment]
@gwwar thanks for reminding me! 🙇 

I've updated this with the support for CPTs.
Unfortunately, the core data doesn't store the CPT labels, so I had to pull the Calypso state back in.

**Comment by Copons** [comment]
> some custom post types like feedback don't have a custom title

It's right there in the PR description! 😏

**Pull request merged**